File: src/packages/flutter-app/lib/models/ops_health.dart
import 'package:json_annotation/json_annotation.dart';

part 'ops_health.g.dart';

/// Dependency + build + backup health from GET /admin/ops/health — an
/// admin-only snapshot powering the System Health dashboard tab. snake_case
/// JSON keys via [FieldRename.snake].

/// Database reachability. [status] is one of "ok" | "unreachable" |
/// "not_configured".
@JsonSerializable(fieldRename: FieldRename.snake)
class HealthDb {
  final String status;
  final int pingMs;

  const HealthDb({this.status = 'not_configured', this.pingMs = 0});

  factory HealthDb.fromJson(Map<String, dynamic> json) =>
      _$HealthDbFromJson(json);
  Map<String, dynamic> toJson() => _$HealthDbToJson(this);
}

/// A single upstream provider and whether its credentials are configured.
@JsonSerializable(fieldRename: FieldRename.snake)
class ProviderStat {
  final String name;
  final bool configured;
  final String note;

  const ProviderStat({
    this.name = '',
    this.configured = false,
    this.note = '',
  });

  factory ProviderStat.fromJson(Map<String, dynamic> json) =>
      _$ProviderStatFromJson(json);
  Map<String, dynamic> toJson() => _$ProviderStatToJson(this);
}

/// Build identity + process uptime.
@JsonSerializable(fieldRename: FieldRename.snake)
class BuildInfo {
  final String release;
  final String goVersion;
  final DateTime? startedAt;
  final int uptimeS;

  const BuildInfo({
    this.release = '',
    this.goVersion = '',
    this.startedAt,
    this.uptimeS = 0,
  });

  factory BuildInfo.fromJson(Map<String, dynamic> json) =>
      _$BuildInfoFromJson(json);
  Map<String, dynamic> toJson() => _$BuildInfoToJson(this);
}

/// Backup freshness. [lastSuccessAt]/[ageS] are null when no backup has been
/// recorded yet; [stale] flags an over-age backup.
@JsonSerializable(fieldRename: FieldRename.snake)
class BackupInfo {
  final String? lastSuccessAt;
  final int? ageS;
  final bool stale;

  const BackupInfo({this.lastSuccessAt, this.ageS, this.stale = false});

  factory BackupInfo.fromJson(Map<String, dynamic> json) =>
      _$BackupInfoFromJson(json);
  Map<String, dynamic> toJson() => _$BackupInfoToJson(this);
}

/// Mirror of GET /admin/ops/health.
@JsonSerializable(fieldRename: FieldRename.snake)
class OpsHealth {
  final HealthDb db;
  final List<ProviderStat> providers;
  final BuildInfo build;
  final BackupInfo backups;
  final bool degraded;
  final List<String> reasons;

  const OpsHealth({
    this.db = const HealthDb(),
    this.providers = const [],
    this.build = const BuildInfo(),
    this.backups = const BackupInfo(),
    this.degraded = false,
    this.reasons = const [],
  });

  factory OpsHealth.fromJson(Map<String, dynamic> json) =>
      _$OpsHealthFromJson(json);
  Map<String, dynamic> toJson() => _$OpsHealthToJson(this);
}

File: src/packages/api/ops_health.go
package main

import (
	"context"
	"net/http"
	"os"
	"runtime"
	"strings"
	"time"
)

// Consolidated dependency-health endpoint (Observability PR2): GET
// /api/v1/admin/ops/health rolls up the DB, every provider's "configured"
// state, build/runtime info, and backup freshness into one DependencyHealth
// document, plus a top-level Degraded verdict + Reasons list.
//
// Like opsMetricsHandler this has NO dbPool guard: the endpoint must render
// even when the database is down — that is exactly when an operator needs it.
// The DB section reports "unreachable"/"not_configured" instead of failing the
// request. Admin auth is still enforced at route registration.
//
// The pure helpers (providerStatuses, evalBackupHealth, computeHealthState) are
// shared verbatim with the self-check monitor (ops_monitor.go) so the endpoint
// and the alert loop can never disagree about what "degraded" means.

const (
	defaultBackupHeartbeatFile = "/opt/goldentempo/backups/.last_success"
	defaultBackupStaleHours    = 36
)

// DependencyHealth is the wire contract of GET /admin/ops/health. Field tags
// are load-bearing: PR3 (Flutter ops dashboard) builds against this exact shape.
type DependencyHealth struct {
	DB        HealthDB       `json:"db"`
	Providers []ProviderStat `json:"providers"`
	Build     BuildInfo      `json:"build"`
	Backups   BackupInfo     `json:"backups"`
	Degraded  bool           `json:"degraded"`
	Reasons   []string       `json:"reasons"` // non-nil (empty slice, not null)
}

// HealthDB reports database reachability. Status is one of "ok",
// "unreachable", or "not_configured"; PingMs is the ping round-trip in ms
// (0 when not ok).
type HealthDB struct {
	Status string `json:"status"`
	PingMs int    `json:"ping_ms"`
}

// ProviderStat is one external dependency's configuration status. Configured
// reflects whether the credential/config needed to use it is present; Note is
// an optional human hint (e.g. "not configured").
type ProviderStat struct {
	Name       string `json:"name"`
	Configured bool   `json:"configured"`
	Note       string `json:"note"`
}

// BuildInfo is static build/runtime provenance for the running process.
type BuildInfo struct {
	Release   string `json:"release"`
	GoVersion string `json:"go_version"`
	StartedAt string `json:"started_at"` // ISO8601 (RFC3339)
	UptimeS   int64  `json:"uptime_s"`
}

// BackupInfo is the freshness of the last successful DB backup, read from the
// heartbeat file the backup job writes. LastSuccessAt/AgeS are nil when no
// heartbeat exists; Stale is true then too.
type BackupInfo struct {
	LastSuccessAt *string `json:"last_success_at"` // ISO8601, null if none
	AgeS          *int64  `json:"age_s"`           // null if none
	Stale         bool    `json:"stale"`
}

// opsHealthHandler serves GET /api/v1/admin/ops/health.
func opsHealthHandler(w http.ResponseWriter, r *http.Request) {
	writeJSON(w, http.StatusOK, buildDependencyHealth(r.Context(), time.Now()))
}

// buildDependencyHealth assembles the full document. now is injectable so the
// backup-freshness math is testable.
func buildDependencyHealth(ctx context.Context, now time.Time) DependencyHealth {
	db := probeDB(ctx)
	backups := readBackupHealth(now)
	state := computeHealthState(db.Status == "ok", backups.Stale)
	return DependencyHealth{
		DB:        db,
		Providers: providerStatuses(),
		Build:     buildInfo(now),
		Backups:   backups,
		Degraded:  state.degraded,
		Reasons:   state.reasons,
	}
}

// probeDB pings the database and reports status + latency. "not_configured"
// when there is no pool, "unreachable" when a ping fails, "ok" otherwise.
func probeDB(ctx context.Context) HealthDB {
	if dbPool == nil {
		return HealthDB{Status: "not_configured"}
	}
	start := time.Now()
	if !pingDB(ctx) {
		return HealthDB{Status: "unreachable"}
	}
	return HealthDB{Status: "ok", PingMs: int(time.Since(start).Milliseconds())}
}

// buildInfo captures static build/runtime provenance.
func buildInfo(now time.Time) BuildInfo {
	return BuildInfo{
		Release:   os.Getenv("SENTRY_RELEASE"),
		GoVersion: runtime.Version(),
		StartedAt: processStart.UTC().Format(time.RFC3339),
		UptimeS:   int64(now.Sub(processStart).Seconds()),
	}
}

// providerStatuses is the consolidated external-dependency configuration list.
// Each entry's Configured flag reuses that provider's own "configured" check so
// this stays the single source of truth for provider readiness. Order is
// stable for a deterministic UI.
func providerStatuses() []ProviderStat {
	stat := func(name string, ok bool) ProviderStat {
		note := ""
		if !ok {
			note = "not configured"
		}
		return ProviderStat{Name: name, Configured: ok, Note: note}
	}

	duffelConfigured := duffelService != nil && duffelService.Token != ""
	emailConfigured := emailService != nil && emailService.Configured()
	transcriptionConfigured := transcriptionService != nil && transcriptionService.Configured()

	return []ProviderStat{
		stat("google_places", os.Getenv("GOOGLE_PLACES_API_KEY") != ""),
		stat("anthropic", os.Getenv("ANTHROPIC_API_KEY") != ""),
		stat("duffel", duffelConfigured),
		stat("ticketmaster", os.Getenv("TICKETMASTER_API_KEY") != ""),
		stat("email", emailConfigured),
		stat("google_oauth", googleOAuthConfigured()),
		stat("transcription", transcriptionConfigured),
		stat("sentry", sentryEnabled),
	}
}

// readBackupHealth reads the backup heartbeat file and evaluates its freshness
// against BACKUP_STALE_HOURS. The file (written by backup.sh) holds a single
// RFC3339 timestamp from `date -u +%FT%TZ`.
func readBackupHealth(now time.Time) BackupInfo {
	path := os.Getenv("BACKUP_HEARTBEAT_FILE")
	if path == "" {
		path = defaultBackupHeartbeatFile
	}
	contents, err := os.ReadFile(path)
	return evalBackupHealth(now, string(contents), err, envInt("BACKUP_STALE_HOURS", defaultBackupStaleHours))
}

// evalBackupHealth is the pure freshness evaluator. A missing/unreadable/
// unparseable heartbeat yields last_success_at=null, age_s=null, stale=true (no
// evidence of a recent backup is itself a degraded signal). Otherwise Stale is
// true once the heartbeat is older than staleHours.
func evalBackupHealth(now time.Time, contents string, readErr error, staleHours int) BackupInfo {
	if readErr != nil {
		return BackupInfo{Stale: true}
	}
	ts, perr := time.Parse(time.RFC3339, strings.TrimSpace(contents))
	if perr != nil {
		return BackupInfo{Stale: true}
	}
	iso := ts.UTC().Format(time.RFC3339)
	ageS := int64(now.Sub(ts).Seconds())
	if ageS < 0 {
		ageS = 0
	}
	stale := now.Sub(ts) > time.Duration(staleHours)*time.Hour
	return BackupInfo{LastSuccessAt: &iso, AgeS: &ageS, Stale: stale}
}

// healthState is the deterministic degraded verdict shared by the endpoint and
// the monitor.
type healthState struct {
	degraded bool
	reasons  []string
}

// computeHealthState derives the top-level degraded verdict from the two
// deterministic, unpriced signals: DB reachability and backup freshness. It
// deliberately does NOT consider provider configuration (a missing optional
// provider key is not "the service is unhealthy") and never pings paid
// providers. reasons is always non-nil.
func computeHealthState(dbOK, backupsStale bool) healthState {
	reasons := []string{}
	if !dbOK {
		reasons = append(reasons, "database unreachable")
	}
	if backupsStale {
		reasons = append(reasons, "backups stale")
	}
	return healthState{degraded: len(reasons) > 0, reasons: reasons}
}
